src/main/java/com/uid2/admin/vertx/service/SharingService.java

```diff
@@ -75,43 +75,81 @@ private void handleSetKeyset(RoutingContext rc) {
+                    || (requestKeysetId != null && requestSiteId != null)) {
                 ResponseUtil.error(rc, 400, "You must specify exactly one of: keyset_id, site_id");
                 return;
             }
 
-            final Map<Integer, Keyset> keysetsById = this.keysetProvider.getSnapshot().getAllKeysets();
-
-            if (keysetId != null) {
-                Keyset keyset = keysetsById.get(keysetId);
+            final int siteId;
+            final int keysetId;
+            final String name;
+            if(requestSiteId != null) {
+                siteId = requestSiteId;
+                name = requestName;
+                if (siteId == Const.Data.AdvertisingTokenSiteId
+                        || siteId == Const.Data.RefreshKeySiteId
+                        || siteId == Const.Data.MasterKeySiteId
+                        || siteProvider.getSite(siteId) == null
+                        || keysetsById.values().stream().anyMatch(k -> k.getSiteId() == siteId))  { // enforce single keyset per site
+                    ResponseUtil.error(rc, 400, "Site id " + siteId + " not valid");
+                    return;
+                }
+                keysetId = Collections.max(keysetsById.keySet()) + 1;
+                if (keysetsById.values().stream().anyMatch(item ->
+                        item.getSiteId() == siteId && item.getName().equalsIgnoreCase(name))) {
+                    ResponseUtil.error(rc, 400, "Keyset with same site_id and name already exists");
+                    return;
+                }
+            } else {
+                Keyset keyset = keysetsById.get(requestKeysetId);
                 if (keyset == null) {
-                    ResponseUtil.error(rc, 404, "Could not find keyset for keyset_id: " + keysetId);
+                    ResponseUtil.error(rc, 404, "Could not find keyset for keyset_id: " + requestKeysetId);
                     return;
                 }
+                keysetId = requestKeysetId;
                 siteId = keyset.getSiteId();
-                if (name.equals("")) {
-                    name = keyset.getName();
+                name = requestName.equals("") ? keyset.getName() : requestName;
+            }
+
+
+            final Set<Integer> newlist;
+
+            if (whitelist != null){
+                Optional<Integer> firstInvalidSite = whitelist.stream().map(s -> (Integer) s).filter(s -> siteProvider.getSite(s) == null).findFirst();
+                if (firstInvalidSite.isPresent()) {
+                    ResponseUtil.error(rc, 400, "Site id " + firstInvalidSite.get() + " not valid");
+                    return;
+                }
+
+                boolean containsDuplicates = whitelist.stream().distinct().count() < whitelist.stream().count();
+                if (containsDuplicates) {
+                    ResponseUtil.error(rc, 400, "Duplicate site_ids not permitted");
+                    return;
                 }
+
+                Integer finalSiteId2 = siteId;
```

Comment:
lets make this explicit that it is final

Reply:
oops, don't need this variable anymore now that siteId is final